src/libraries/System.IO.Compression/src/System/IO/Compression/WinZipAesStream.cs

```diff
@@ -0,0 +1,770 @@
+        {
+            ArgumentNullException.ThrowIfNull(baseStream);
+
+
+            _baseStream = baseStream;
+            _password = password;
+            _encrypting = encrypting;
+            _keySizeBits = keySizeBits;
+            _totalStreamSize = totalStreamSize; // Store the total size
+            _bytesReadFromBase = 0;
+            _leaveOpen = leaveOpen;
+#pragma warning disable CA1416 // HMACSHA1 is available on all platforms
+            _aes = Aes.Create();
+#pragma warning restore CA1416
+            _aes.Mode = CipherMode.ECB;
+            _aes.Padding = PaddingMode.None;
+
+#pragma warning disable CA1416 // HMACSHA1 available on all platforms ?
+#pragma warning disable CA5350 // Do Not Use Weak Cryptographic Algorithms ?
+            _hmac = new HMACSHA1();
+#pragma warning restore CA5350 // Do Not Use Weak Cryptographic Algorithms
+#pragma warning restore CA1416
+
+            Array.Clear(_counterBlock, 0, 16);
+            _counterBlock[0] = 1;
+
+            if (_encrypting)
+            {
+                _encryptionBuffer = new MemoryStream();
+                GenerateKeys();
+                InitCipher();
+            }
+        }
+
+        private void DeriveKeysFromPassword()
+        {
+            Debug.Assert(_salt is not null, "Salt must be initialized before deriving keys");
+
+            byte[] passwordBytes = Encoding.UTF8.GetBytes(_password.ToArray());
+
+            try
+            {
+                // AES key size + HMAC key size (same as AES key) + password verifier (2 bytes)
+                int keySizeInBytes = _keySizeBits / 8;
+                int totalKeySize = keySizeInBytes + keySizeInBytes + 2;
+
+                // WinZip AES uses SHA1 for PBKDF2 with 1000 iterations per spec
+                byte[] derivedKey = Rfc2898DeriveBytes.Pbkdf2(
+                passwordBytes,
+                _salt!,
+                1000,
+                HashAlgorithmName.SHA1,
+                totalKeySize);
+
+                // Split the derived key material
+                _key = new byte[keySizeInBytes];
+                _hmacKey = new byte[keySizeInBytes];
+                _passwordVerifier = new byte[2];
+
+                // Copy the key material in the correct order
+                int offset = 0;
+
+                // First: AES encryption key
+                Buffer.BlockCopy(derivedKey, offset, _key, 0, _key.Length);
+                offset += _key.Length;
+
+                // Second: HMAC authentication key (same size as encryption key)
+                Buffer.BlockCopy(derivedKey, offset, _hmacKey, 0, _hmacKey.Length);
+                offset += _hmacKey.Length;
+
+                // Third: Password verification value (2 bytes)
+                Buffer.BlockCopy(derivedKey, offset, _passwordVerifier, 0, _passwordVerifier.Length);
```

Comment:
prefer more modern/flexible Span-based APIs 
```suggestion
                // First: AES encryption key
                derivedKey.AsSpan(0, _key.Length).CopyTo(_key);

                // Second: HMAC authentication key (same size as encryption key)
                derivedKey.AsSpan(_key.Length, _hmacKey.Length).CopyTo(_hmacKey);

                // Third: Password verification value (2 bytes)
                derivedKey.AsSpan(_key.Length + _hmacKey.Length).CopyTo( _passwordVerifier);
```